### PR TITLE
device: supported devices API

### PR DESCRIPTION
**Pull request by JordanYates**
Implements an API to iterate over devicetree devices that an arbitrary device supports.
Follows the conventions setup by the dependent devices API.

Includes a minor optimization to the handles array (saves one handle per array), which offsets the cost of the additional `DEVICE_HANDLE_SEP`. Therefore the ROM increase is purely a function of the number of supported devices per device (2 bytes per).

Implements #37793.

### CODE REVIEW

**Comment by ceolin** [comment]
cool, I have the same feature implemented in a different pr https://github.com/zephyrproject-rtos/zephyr/pull/37723. Lets me try my feature on top of your supported devices implementation. I'm prone to your implementation, it is looking cleaner than mine.

**Review by mbolivar-nordic — COMMENTED** [review]
Can you please split the functional changes from the whitespace changes?

**Comment by JordanYates** [comment]
> Can you please split the functional changes from the whitespace changes?

Done

**Comment by JordanYates** [comment]
> cool, I have the same feature implemented in a different pr #37723. 

I hadn't realized there was already an implementation, or I wouldn't have bothered to try it myself.
I do think adding the handles to the existing array is a better solution than a new array.

**Comment by ceolin** [comment]
> > cool, I have the same feature implemented in a different pr #37723.
> 
> I hadn't realized there was already an implementation, or I wouldn't have bothered to try it myself.
> I do think adding the handles to the existing array is a better solution than a new array.

Agree, as soon as this pr gets in, I'll rebase that one.

**Review by ceolin — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: include/device.h [hunk 1]
```diff
@@ -487,6 +487,46 @@ device_required_handles_get(const struct device *dev,
 	return rv;
 }
 
+/**
+ * @brief Get the set of handles for devicetree supports of this device.
```
Comment (nashif) [inline include/device.h:491]: can you please rephrase that?

File: include/device.h [hunk 2]
```diff
@@ -487,6 +487,46 @@ device_required_handles_get(const struct device *dev,
...
+ *
+ * These are the device supports inferred from devicetree.
```
Comment (nashif) [inline include/device.h:493]: same here, what do "device supports" and "devicetree supports" mean in this context?

File: include/device.h [hunk 3]
```diff
@@ -524,6 +564,42 @@ int device_required_foreach(const struct device *dev,
...
+ * @param visitor_cb the function that should be invoked on each device in the
+ * support set. This parameter must not be null.
+ *
+ * @param context state that is passed through to the visitor function. This
+ * parameter may be null if @p visitor tolerates a null @p context.
+ *
+ * @return The number of devices that were visited if all visits succeed, or
+ * the negative value returned from the first visit that did not succeed.
+ */
+int device_supported_foreach(const struct device *dev,
```
Comment (nashif) [inline include/device.h:599]: I am wondering if "supported" is the right language we want to use here, maybe "connected"? Then the @brief would be "Visit every device that is directly connected to @p dev."

**Review by nashif — CHANGES_REQUESTED** [review]

File: include/device.h (hunk 3 above)
Comment (JordanYates) [reply include/device.h:599]: Supports is the terminology used in the existing API's. I do agree it makes for awkward documentation.
https://github.com/zephyrproject-rtos/zephyr/blob/cc760c54018ba142d283350e653881b11485c4ff/include/devicetree/ordinals.h#L47-L61

File: include/device.h (hunk 1 above)
Comment (JordanYates) [reply include/device.h]: Let me know if the update is clearer

File: include/device.h (hunk 3 above)
Comment (nashif) [reply include/device.h:600]: > Supports is the terminology used in the existing API's. I do agree it makes for awkward documentation.

yes it looks very awkward for someone not familiar with the DT APIs. Is this a zephyr thing or does this come from the devicetree spec? @mbolivar-nordic @galak

File: include/device.h (hunk 1 above)
Comment (nashif) [reply include/device.h]: yes, looks much better.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**